### ISSUE TITLE
Create dataset, loading and pre-processing utilities

### ISSUE THREAD

**Issue by martinnormark** [martinnormark/neural-mesh-simplification#2]
- [x] Implement functions to load and preprocess mesh data.
- [x] Implement efficient data loading for training.

Based on the project structure provided, you should implement the data loading and preprocessing functionalities primarily in the `data/` directory. Specifically, you'll want to focus on the `data/dataset.py` file. Here's a breakdown of where to put different components and what functions you might need:

1. In `data/dataset.py`:

This file should contain your main dataset class and related utility functions. Here's an outline of functions you might include:

```python
import torch
from torch.utils.data import Dataset
import trimesh

class MeshSimplificationDataset(Dataset):
    def __init__(self, data_dir, transform=None):
        # Initialize dataset
        pass

    def __len__(self):
        # Return length of dataset
        pass

    def __getitem__(self, idx):
        # Load and return a single data sample
        pass

def load_mesh(file_path):
    # Load a mesh from file
    pass

def preprocess_mesh(mesh):
    # Preprocess a mesh (e.g., normalize, center)
    pass

def augment_mesh(mesh):
    # Apply data augmentation to a mesh
    pass

def mesh_to_tensor(mesh):
    # Convert a mesh to tensor representation
    pass
```

2. In `utils/mesh_operations.py`:

This file can contain more general mesh manipulation functions that might be used outside of just data loading:

```python
def simplify_mesh(mesh, target_faces):
    # Simplify a mesh to a target number of faces
    pass

def calculate_mesh_features(mesh):
    # Calculate relevant features of a mesh (e.g., curvature)
    pass

def align_meshes(mesh1, mesh2):
    # Align two meshes (useful for comparison)
    pass
```

3. In `scripts/preprocess_data.py`:

You might want to create a new script for batch preprocessing of your data:

```python
import os
from data.dataset import load_mesh, preprocess_mesh
from utils.mesh_operations import simplify_mesh

def preprocess_dataset(input_dir, output_dir, target_faces):
    # Iterate through all meshes in input_dir
    # Load, preprocess, simplify, and save to output_dir
    pass

if __name__ == "__main__":
    preprocess_dataset("data/raw", "data/processed", target_faces=1000)
```

4. Additional considerations:

- You might need to install additional libraries like `trimesh` or `open3d` for mesh processing. Add these to your `requirements.txt`.
- Consider implementing a custom `collate_fn` in `data/dataset.py` if you're dealing with meshes of varying sizes.
- You may want to add caching mechanisms to speed up data loading, especially if preprocessing is time-consuming.

Here's a list of key functions you'll need to implement:

1. `load_mesh(file_path)`: Load a mesh from various file formats.
2. `preprocess_mesh(mesh)`: Normalize, center, and prepare the mesh for the network.
3. `augment_mesh(mesh)`: Apply data augmentation techniques (e.g., rotation, scaling).
4. `mesh_to_tensor(mesh)`: Convert mesh data to tensor format for PyTorch.
5. `simplify_mesh(mesh, target_faces)`: Simplify a mesh to a target number of faces.
6. `calculate_mesh_features(mesh)`: Extract relevant features from the mesh.
7. `align_meshes(mesh1, mesh2)`: Align two meshes for comparison.
8. `batch_process_meshes(input_dir, output_dir)`: Process all meshes in a directory.

Remember to handle different mesh formats, deal with potential errors in the meshes, and ensure your preprocessing steps are consistent across all your data. Also, consider implementing parallel processing for faster data preparation, especially if you're dealing with a large dataset.
